lib/collection/src/shards/transfer/wal_delta.rs

```diff
@@ -0,0 +1,174 @@
+///   shard participate on all nodes.
+///
+/// After this function, the following will happen:
+///
+/// - The local shard is un-proxified
+/// - The shard transfer is finished
+/// - The remote shard state is set to `Active` through consensus
+///
+/// # Cancel safety
+///
+/// This function is cancel safe.
+///
+/// If cancelled - the remote shard may only be partially recovered/transferred and the local shard
+/// may be left in an unexpected state. This must be resolved manually in case of cancellation.
+pub(super) async fn transfer_wal_delta(
+    transfer_config: ShardTransfer,
+    shard_holder: Arc<LockedShardHolder>,
+    shard_id: ShardId,
+    remote_shard: RemoteShard,
+    channel_service: ChannelService,
+    consensus: &dyn ShardTransferConsensus,
+    collection_name: &str,
+) -> CollectionResult<()> {
+    let remote_peer_id = remote_shard.peer_id;
+
+    log::debug!("Starting shard {shard_id} transfer to peer {remote_peer_id} using diff transfer");
+
+    // Ask remote shard on failed node for recovery point
+    let recovery_point = remote_shard
+        .shard_recovery_point(collection_name, shard_id)
+        .await
+        .map_err(|err| {
+            CollectionError::service_error(format!(
+                "Failed to request recovery point from remote shard: {err}"
+            ))
+        })?;
+
+    let shard_holder_read = shard_holder.read().await;
+
+    let transferring_shard = shard_holder_read.get_shard(&shard_id);
+    let Some(replica_set) = transferring_shard else {
+        return Err(CollectionError::service_error(format!(
+            "Shard {shard_id} cannot be queue proxied because it does not exist"
+        )));
+    };
+
+    // Resolve WAL delta, get the version to start the diff from
+    let wal_delta_version = replica_set
+        .resolve_wal_delta(recovery_point)
+        .await
+        .map_err(|err| {
+            CollectionError::service_error(format!("Failed to resolve shard diff: {err}"))
+        })?;
+
+    if let Some(wal_delta_version) = wal_delta_version {
+        // Queue proxy local shard
+        replica_set
+            .queue_proxify_local(remote_shard.clone(), Some(wal_delta_version))
+            .await?;
+
+        debug_assert!(
+            replica_set.is_queue_proxy().await,
+            "Local shard must be a queue proxy",
+        );
+
+        // Transfer queued updates to remote, transform into forward proxy
+        // This way we send a complete WAL diff
+        log::trace!("Transfer WAL diff by transferring all queue proxy updates and transform into forward proxy");
+        replica_set.queue_proxy_into_forward_proxy().await?;
+
+        // Update cutoff point on remote shard, disallow recovery before our current last seen
+        let cutoff = replica_set.shard_recovery_point().await?;
+        remote_shard
+            .update_shard_cutoff_point(collection_name, shard_id, &cutoff)
+            .await?;
```

Comment:
I don't think we need this for diff transfer?

Like, imagine you just transferred a diff of exactly one point. So, recovered node now has more-or-less exactly the same WAL as the current node. It will be able to resolve any diff just fine. Why forcibly update the cutoff?

Reply:
I think you're right.

Before we had this issue with ordering during the WAL delta transfer. But we basically resolved that now with an extra replica set state. I'll remove it.